pkg/skaffold/runner/dev.go

```diff
@@ -33,36 +33,64 @@ import (
 // ErrorConfigurationChanged is a special error that's returned when the skaffold configuration was changed.
 var ErrorConfigurationChanged = errors.New("configuration changed")
 
-func (r *SkaffoldRunner) doDev(ctx context.Context, out io.Writer) error {
-	defer r.changeSet.reset()
+func (r *SkaffoldRunner) needsAction() bool {
+	// NOTE: r.intentLock must be acquired before calling this function!
+	return (r.changeSet.needsRedeploy && r.deployIntent) ||
+		(len(r.changeSet.needsRebuild) > 0 && r.buildIntent) ||
+		(len(r.changeSet.needsResync) > 0 && r.syncIntent)
+}
 
+func (r *SkaffoldRunner) doDev(ctx context.Context, out io.Writer) error {
 	r.logger.Mute()
 
-	if r.changeSet.needsAction() {
+	r.intentLock.Lock()
+	defer r.intentLock.Unlock()
```

Comment:
we should make the lock finer grained...
```
r.intentLock.Lock()
...
syncIntent := r.syncIntent
if len(r.changeSet.needsResync) > 0 && r.syncIntent { 
 r.resetSyncIntent()
}
...
r.intentLock.Unlock()
```

Reply:
done